src/hotspot/share/gc/g1/g1EvacFailure.cpp

```diff
@@ -29,171 +29,230 @@
-    obj->init_mark();
+    _start(Ticks::now()) { }
 
-    HeapWord* obj_end = obj_addr + obj_size;
-    _last_forwarded_object_end = obj_end;
-    _hr->update_bot_for_block(obj_addr, obj_end);
-    return obj_size;
+  ~PhaseTimesStat() {
+    _phase_times->record_or_add_time_secs(phase_name,
+                                          _worker_id,
+                                          (Ticks::now() - _start).seconds());
   }
 
-  // Fill the memory area from start to end with filler objects, and update the BOT
-  // accordingly.
-  void zap_dead_objects(HeapWord* start, HeapWord* end) {
-    if (start == end) {
-      return;
-    }
+  void register_empty_chunk() {
+    _phase_times->record_or_add_thread_work_item(phase_name,
+                                                 _worker_id,
+                                                 1,
+                                                 G1GCPhaseTimes::RemoveSelfForwardEmptyChunksNum);
+  }
 
-    _hr->fill_range_with_dead_objects(start, end);
+  void register_nonempty_chunk() {
+    _phase_times->record_or_add_thread_work_item(phase_name,
+                                                 _worker_id,
+                                                 1,
+                                                 G1GCPhaseTimes::RemoveSelfForwardChunksNum);
   }
 
-  void zap_remainder() {
-    zap_dead_objects(_last_forwarded_object_end, _hr->top());
+  void register_objects_count_and_size(size_t num_marked_obj, size_t marked_words) {
+    _phase_times->record_or_add_thread_work_item(phase_name,
+                                                 _worker_id,
+                                                 num_marked_obj,
+                                                 G1GCPhaseTimes::RemoveSelfForwardObjectsNum);
+
+    size_t marked_bytes = marked_words * HeapWordSize;
+    _phase_times->record_or_add_thread_work_item(phase_name,
+                                                 _worker_id,
+                                                 marked_bytes,
+                                                 G1GCPhaseTimes::RemoveSelfForwardObjectsBytes);
   }
 };
 
-class RemoveSelfForwardPtrHRClosure: public HeapRegionClosure {
+// Fill the memory area from start to end with filler objects, and update the BOT
+// accordingly. Since we clear and use the bitmap for marking objects that failed
+// evacuation, there is no other work to be done there.
+static size_t zap_dead_objects(HeapRegion* hr, HeapWord* start, HeapWord* end) {
+  assert(start <= end, "precondition");
+  if (start == end) {
+    return 0;
+  }
+
+  hr->fill_range_with_dead_objects(start, end);
+  return pointer_delta(end, start);
+}
+
+static void prefetch_obj(HeapWord* obj_addr) {
+  Prefetch::write(obj_addr, PrefetchScanIntervalInBytes);
+}
+
+// Caches the currently accumulated number of garbage words found in this heap region.
+// Avoids direct (frequent) atomic operations on the HeapRegion's garbage counter.
```

Comment:
I think that's it. We expect few failed regions (1-2) with pinned regions, so many threads working on the same region at the same time. If you want we can drop the cache, I did not (re-)measure.

Reply:
I would prefer to drop it if not needed. From the comment it sounds like the cache was created to avoid updating the global value for each "garbage area", but a middle road would be to update the global value after each chunk. How much work is it to re-measure?

Reply:
I intend to work and measure in this area again soon'ish so I'll have plenty of opportunity to see if makes a difference. For this change I'll remove the cache.